src/DistributedArrays.jl

```diff
@@ -1143,24 +1143,42 @@ end
 
 
 # Level 3
-function A_mul_B!(α::Number, A::DMatrix, B::AbstractMatrix, β::Number, C::DMatrix)
-
+function _matmatmul!(α::Number, A::DMatrix, B::AbstractMatrix, β::Number, C::DMatrix, tA)
     # error checks
-    if size(A, 2) != size(B, 1)
-        throw(DimensionMismatch(""))
+    Ad1, Ad2 = (tA == 'N') ? (1,2) : (2,1)
+    mA, nA = size(A, Ad1, Ad2)
+    mB, nB = size(B)
+    if mB != nA
+        throw(DimensionMismatch("matrix A has dimensions ($mA, $nA), matrix B has dimensions ($mB, $nB)"))
+    end
+    if size(C,1) != mA || size(C,2) != nB
+        throw(DimensionMismatch("result C has dimensions $(size(C)), needs ($mA, $nB)"))
     end
-    if C.cuts[1] != A.cuts[1]
-        throw(ArgumentError("cuts of the first dimension of the output matrix must match cuts of first dimension of the first input matrix"))
+    if C.cuts[1] != A.cuts[Ad1]
+        throw(ArgumentError("cuts of the first dimension of the output matrix must match cuts of dimension $Ad1 of the first input matrix"))
     end
 
     # Multiply on each tile of A
-    R = Array(Future, size(procs(A))..., size(procs(C), 2))
-    for j = 1:size(A.pids, 2)
+    if tA == 'N'
+        R = Array(Future, size(procs(A))..., size(procs(C), 2))
+    else
+        R = Array(Future, reverse(size(procs(A)))..., size(procs(C), 2))
+    end
+    for j = 1:size(A.pids, Ad2)
         for k = 1:size(C.pids, 2)
-            Bjk = B[A.cuts[2][j]:A.cuts[2][j + 1] - 1, C.cuts[2][k]:C.cuts[2][k + 1] - 1]
-            for i = 1:size(A.pids, 1)
-                R[i,j,k] = remotecall(procs(A)[i,j]) do
-                    localpart(A)*convert(localtype(B), Bjk)
+            Acuts = A.cuts[Ad2]
+            Ccuts = C.cuts[2]
+            Bjk = B[Acuts[j]:Acuts[j + 1] - 1, Ccuts[k]:Ccuts[k + 1] - 1]
+            for i = 1:size(A.pids, Ad1)
+                p = (tA == 'N') ? procs(A)[i,j] : procs(A)[j,i]
+                R[i,j,k] = remotecall(p) do
+                    Atc = localpart(A)
+                    if tA == 'T'
+                        Atc = transpose(Atc)
```

Comment:
The reason is that
```jl
julia> expand(:(A'*B))
:(Ac_mul_B(A,B))      
```
and `Ac_mul_B` eventually calls [`xgemm`](http://www.netlib.org/lapack/explore-html/d1/d54/group__double__blas__level3.html#gaeda3cbd99c8fb834a60a6412878226e1) which computes `C:=α*A'B + β*C` by updating `C` in place such that `A` is never transposed.